core/src/main/java/org/apache/calcite/rel/type/TimeFrameSet.java

```diff
@@ -0,0 +1,762 @@
+          new TimeFrameSet(ImmutableMap.copyOf(map), rollupList.build());
+      frameSetSupplier.accept(frameSet);
+      return frameSet;
+    }
+
+    /** Converts a number to an exactly equivalent {@code BigInteger}.
+     * May silently lose precision if n is a {@code Float} or {@code Double}. */
+    static BigInteger toBigInteger(Number number) {
+      return number instanceof BigInteger ? (BigInteger) number
+          : BigInteger.valueOf(number.longValue());
+    }
+
+    public Builder addCore(String name) {
+      map.put(name, new CoreFrame(frameSetSupplier, name));
+      return this;
+    }
+
+    /** Defines a time unit that consists of {@code count} instances of
+     * {@code baseUnit}. */
+    Builder addSub(String name, boolean divide, Number count,
+        String baseName, TimestampString epoch) {
+      final TimeFrameImpl baseFrame = get(baseName);
+      final BigInteger factor = toBigInteger(count);
+
+      final CoreFrame coreFrame = baseFrame.core();
+      final BigFraction coreFactor = divide
+          ? baseFrame.coreMultiplier().divide(factor)
+          : baseFrame.coreMultiplier().multiply(factor);
+
+      map.put(name,
+          new SubFrame(name, baseFrame, divide, factor, coreFrame,
+              coreFactor, epoch));
+      return this;
+    }
+
+    /** Defines a time unit that is the number of a minor unit within a major
+     * unit. For example, the "DOY" frame has minor "DAY" and major "YEAR". */
+    Builder addQuotient(String name, String minorName, String majorName) {
+      final TimeFrameImpl minorFrame = get(minorName);
+      final TimeFrameImpl majorFrame = get(majorName);
+      map.put(name, new QuotientFrame(name, minorFrame, majorFrame));
+      return this;
+    }
+
+    /** Returns the time frame with the given name,
+     * or throws {@link IllegalArgumentException}. */
+    TimeFrameImpl get(String name) {
+      final TimeFrameImpl timeFrame = map.get(name);
+      if (timeFrame == null) {
+        throw new IllegalArgumentException("unknown frame: " + name);
+      }
+      return timeFrame;
+    }
+
+    /** Defines a time unit that consists of {@code count} instances of
+     * {@code baseUnit}. */
+    public Builder addMultiple(String name, Number count,
+        String baseName) {
+      return addSub(name, false, count, baseName, TimestampString.EPOCH);
+    }
+
+    /** Defines such that each {@code baseUnit} consists of {@code count}
+     * instances of the new unit. */
+    public Builder addDivision(String name, Number count, String baseName) {
+      return addSub(name, true, count, baseName, TimestampString.EPOCH);
+    }
+
+    /** Defines a rollup from one frame to another.
+     *
+     * <p>An explicit rollup is not necessary for frames where one is a multiple
+     * of another (such as MILLISECOND to HOUR). Only use this method for frames
+     * that are not multiples (such as DAY to MONTH). */
+    public Builder addRollup(String fromName, String toName) {
```

Comment:
Would all custom timeframes require explicit rollups?

Reply:
No, we can infer that you can rollup from, say, `minute15` to `hour7` because we know that one is 15 * 60 seconds and the other is 7 * 60 * 60 seconds.

A month is not a fixed multiple of days. An explicit rollup is basically saying that the start of a month is aligned with the start of a day. And therefore you can compute a month total by adding up the day totals for all days in that month.

The same is not true for weeks in (non-ISO) years. A year does not start on the first day of a week. You cannot roll up weeks to years.

Good question, though. I'll add an explanation in the javadoc.